### ISSUE TITLE
Code Insights: Adjust label chart layouts for particular insight charts

### ISSUE THREAD

**Issue by vovakulikov** [sourcegraph/sourcegraph#39935]
Follow up for https://github.com/sourcegraph/sourcegraph/pull/39805

## Background

In https://github.com/sourcegraph/sourcegraph/pull/39805 we added a few new options to how we can position axis labels on the chart. See https://github.com/sourcegraph/sourcegraph/pull/39805#issuecomment-1204030258

In this issue we should decide how exactly we want to apply these options for compute powered and aggregations on the search page insights charts. 

From @felixfbecker in https://github.com/sourcegraph/sourcegraph/pull/39805
> To me it makes more sense to omit bars and labels at the tail end (low numbers = less important), than to hide the labels for every second bar, because they are usually sorted in descending importance. So seeing the label for bar 2 and 4 is very important, but would be hidden in approach (3).
>
> I also think we should consider not showing the legend by default. I don't see much value from it for the space it takes up, considering it duplicates text from the labels and only shows a few points above the fold. The legend usually describes the different series names, but here weirdly it describes each data point (it's a list of each X value), which is unusual. For 4.0, we want to disable multiple series for group insight types, so the legend isn't necessary even if it was showing series.
>
>Hiding the legend may solve the problem already if we keep the number of bars the same.

/cc @joelkw @felixfbecker @vovakulikov @unclejustin

**Comment by vovakulikov** [comment]
Feedback from @valerybugakov (https://github.com/sourcegraph/sourcegraph/pull/39805#pullrequestreview-1064730060)
>As a consumer who visited compute charts multiple times, I had a similar feeling, so I like this proposal. It often felt that we were trying to display too much data on a limited surface. Making this behavior responsive to the available space would be an ideal solution for me — if space between bars goes under a certain threshold — show fewer bars.

**Comment by Joelkw** [comment]
Notes from sync: 
- This will be split into 2 issues: 
  - 1 completed we have different logic for different chart types
  - 1 backlogged for grouping other categories on the chart

**Comment by vovakulikov** [comment]
Closing this since it was closed for aggregation project.

**Comment by felixfbecker** [comment]
@vovakulikov I don't think this was solved in that PR. For aggregations, I still see "skipped" labels, rather than always labelling the first X bars and showing no labels for the tail.

<img width="241" alt="image" src="https://user-images.githubusercontent.com/10532611/188010575-ab0e4622-7320-4cf5-a3b9-51b3496efe87.png">

**Comment by felixfbecker** [comment]
I don't think it _blocks_ 4.0, should we propose it for iteration 26?

**Comment by vovakulikov** [comment]
@felixfbecker, sorry this issue was originally created for the different truncation logic for the labels. This is why I closed this one (but this was a wrong call since this issue has evolved since then, thanks for spotting this) 

@AlicjaSuska, I think this issue requires your attention. In Figma designs for the aggregation UI, we have the following design. 

<img width="417" alt="Screenshot 2022-09-05 at 10 09 18" src="https://user-images.githubusercontent.com/18492575/188386113-1e2ca102-90d7-4b47-9848-07a2798a8645.png">

This matches with the current statement on this issue 
>To me it makes more sense to omit bars and labels at the tail end (low numbers = less important), than to hide the labels for every second bar, because they are usually sorted in descending importance. So seeing the label for bar 2 and 4 is very important, but would be hidden in approach (3).

But I see that this wouldn't work for the aggregation UI for all cases. The design represents a happy path scenario where labels at the Y axis don't spread chart content too much. For big values and hence long labels at Y axis, we would have less space for the chart bars, this means we make each bar chart thinner; therefore, we have less space between labels at the X axis (even if we don't render labels at the end, because all bars have equal width we have the equal space for all labels regardless of their position on the chart) 

In this case, we should hide the lower bars from the chart completely and, by this, give more space for the bars at the beginning. But this may be non-optimal cause we render less information than we have.

**Comment by AlicjaSuska** [comment]
@vovakulikov I'm not sure I'm following your train of thought:

> For big values and hence long labels at Y axis, we would have less space for the chart bars, this means we make each bar chart thinner

This design represents the maximum number of chart bars and their minimum width - we can't display more and thinner bars because of the accessibility issues (click target would be too small)

What I understood from @felixfbecker 's initial comment and then this follow-up:

> I still see "skipped" labels, rather than always labelling the first X bars and showing no labels for the tail.

is that he suggests that we change the logic of which labels are displayed (example: chart shown above with 10 bars and 5 labels)
- from: displaying every other label (currently every other bar has a line below it that indicates that it is labeled) - bars 1,3,5,7,9 are labeled
- to: displaying labels starting from the highest value bars and skipping the labels at the end of the chart - bards 1-5 would be labeled

In general, I see the logic behind it and I agree that there is more value in showing labels for the higher value bars. 

Here is how it would look like if we had pointers for the first 5 bars and labels below them. It's impossible to align them with the pointers and it may be confusing for the users when they try to read which label is describing which bar. @felixfbecker @vovakulikov wdyt about how it looks like? Is it confusing for you or the fact that some bars have pointers is enough of an indication?
<img width="613" alt="Screenshot 2022-09-05 at 11 41 57" src="https://user-images.githubusercontent.com/20326070/188419890-fa6eb134-6e72-4b63-9729-420131de423d.png">

**Comment by vovakulikov** [comment]
@AlicjaSuska I tried to draw this example about Y long labels. You see how Y labels push chart content and make bars thinner. Because of this X labels become closer to each other and it becomes hard to read

![Untitled-2022-08-29-1824-2](https://user-images.githubusercontent.com/18492575/188425899-4d8f1148-32ce-46d2-bb7a-d7c61b150602.png)

**Comment by AlicjaSuska** [comment]
Thank you for this drawing @vovakulikov - I understand what you mean now. In this case, I think we should just display fewer bars that still meet the accessibility standards for clickable targets. We can't make the y axis labels text any smaller - we're also at the limit there.

**Comment by AlicjaSuska** [comment]
@vovakulikov I think we're a little lost in the discussion here. We have 2 separate problems:
- the y axis labels pushing the bars (that you talk about)
- the x axis labels - which to display (that Felix and I were talking about)

Just wanted to clarify this to make sure we understand each other well.

**Comment by vovakulikov** [comment]
@AlicjaSuska I agree these are kind of two separate problems but the first one in my opinion affects the second one. 

Just to make on the same page here - anything about labels layout now should be considered as follow up work (unless it's something simple and small) because logic there isn't so straightforward and considering a given timeline and code freeze date I would like to avoid any big changes about labels and charts. 

So I propose to move this out of 4.0 release and mark this one as follow up for post 4.0 release.
